### PR TITLE
Register Test Contract

### PR DESCRIPTION
**Pull request by jonjove**
### What

Add machinery to register a test contract and call it.

### Why

Testing multiple contracts is basically impossible right now. This facilitates calling contracts by contract identifier.

### Known limitations

The interface isn't amazing.

### CODE REVIEW

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:181]: I understand what a vtable, but I don't think it should be a concept exposed by the SDK. What does a `ContractVTableFnTrait` represent? Does it represent a function that can be called with a list of arguments?
```suggestion
pub trait ContractFn {
```

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:932]:
```suggestion
    pub fn register_contract(
```

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:184]: What is `duplicate`'s behavior, and how is it different to the behavior of `clone`?

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [inline stellar-contract-env-host/src/host.rs:792]: What's the intended future behavior of this fn when not in test mode, it seems odd to always return None.

**Review by leighmcculloch — COMMENTED** [review]
👏🏻 I definitely think we need something like this.

Could you provide an example of how this will be used? It isn't clear how all components of interface will be used, like the try_ and call_ fns, and I suspect we should at least rename some of these things.

Some comments inline.

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [reply stellar-contract-env-host/src/host.rs:184]: Actually, why do you need to be able to clone fns? That seems surprising.

File: stellar-contract-env-host/src/host.rs
Comment (jayz22) [inline stellar-contract-env-host/src/host.rs:860]: Nitpick: I don't feel the vm call should be "nested" inside the vtable. They are more like parallel relationship instead of one being the fallback of the other (which `or_else` implies). 
If they are nested, then if feels more appropriate the other way around (`try_call_vm.or_else(try_call_vtable)`)? Since VM is the more default/production focused branch.

Comment (leighmcculloch) [reply stellar-contract-env-host/src/host.rs:860]: I think once we have native contracts the relationship might exist as such where we execute a vtable loaded contract instead of a vm one if it is present. In tests we are really just presenting the test contract as a native contract.

Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:860]: I wouldn't call this nested, it's sequential. That is the appropriate relationship because I can only call one function so there must exist an order in which they are considered. This `or_else` is just a very compact expression of this relationship. As @leighmcculloch points out, if native functions were here then the order would be "test, native, vm".

But I disagree that it should be the other way around. If I've registered a test contract, I expect it to get called so I can debug appropriately.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:184]: This goes a couple of layers deep. The root cause is that `EnvBase` supports `deep_clone`, which facilitates step-forward/step-backward state checking. For `Host`, this is implemented in terms of `HostImpl::clone`. Therefore `HostImpl` is `Clone`, which means that all parts of `HostImpl` need to be `Clone` as well. Specifically, `RefCell<std::collections::HashMap<Hash, ContractVTable>>` needs to be `Clone`, which is the case only if `ContractVTable` is `Clone`, which is the case only if `ContractVTableFn` is `Clone`. In summary, if we want to change this, then we need to change how step-forward/step-backward state checking works.

So how is `duplicate` different from `clone`? `Box<dyn Clone>` is not [object safe](https://doc.rust-lang.org/reference/items/traits.html#object-safety) because `Clone` uses `Self` outside of the receiver. This `duplicate` machinery, in contrast, is object safe. It works in conjunction with `Box<...>` by constructing a new `Box<...>` by cloning the boxed object.

Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:184]: It might be possible to provide a custom implementation of `Clone` for `std::collections::HashMap<Hash, ContractVTable>` using the newtype idiom. I don't know if that will be better, but I will investigate now.

Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: It took a bit of finagling, but I think I was able to get this cleaned up. Is it better? Maybe. Did it allow me to find a different design flaw and fix it? Yes.

We can do more work on this interface in the future, but I don't want to block this PR on getting the best possible interface right now. It's important to remember that my goal is for this machinery to end up behind a macro. I'm already partially doing that in rs-stellar-token-contract.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: I was able to replace with `Fn(&[RawVal]) -> RawVal` (note: _not_ `Fn(RawVal) -> RawVal` because it must accept a variable number of parameters), although it took some work. The trait no longer exists so no name to change.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: What is a contract in this context, other than a vtable?

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: I reworked this to avoid this quirk.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: I renamed this `call_wasm_fn` to make it more obvious what it actually does. It's just a helper function for `call_n` to avoid it being ridiculously long.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: For the record, I removed this code but the new behavior is semantically very similar. The only difference is that in the final case it returns a host error instead of panic with `unimplemented`.

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs:894]: I don't want to make this change. This machinery is explicitly for test contracts. Native contracts cannot work this way.

**Comment by jonjove** [comment]
> Could you provide an example of how this will be used?

I have a complete example using this machinery in rs-stellar-token-contract. Once this gets merged, I'll open the small SDK PR. Once that gets merged I'll open the token example.

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [reply stellar-contract-env-host/src/host.rs]: I don't think we know the answer to that yet.

If you prefer to keep the narrower focus I suggest `ContractFunctionSet`. I suggest that because if a developer asks me what the ContractVTable is, not knowing what a vtable is, I would tell them it is the functions of the contract.

I mostly want to remove the term vtable from the public API of the SDK. So even if it was only a vtable, I don't think we need to surface the term vtable. This is part of the API that contract developers will use. The term vtable is only common in certain technology ecosystems and it is not meaningful to probably a large variety of developers. I hope we can wrap this inside a macro, but even then I think the term vtable is unnecessary. It's simple to say that this is a `Contract`, or that this is a `ContractFunctionSet`.

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [reply stellar-contract-env-host/src/host.rs]: Is a commit missed from being pushed? This code is still present in the PR.

Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: I no longer see this code. Did you refresh?

File: stellar-contract-env-host/src/host.rs
Comment (jonjove) [reply stellar-contract-env-host/src/host.rs]: Done.

File: stellar-contract-env-host/src/host.rs
Comment (leighmcculloch) [reply stellar-contract-env-host/src/host.rs:938]: 👍🏻 Let's circle back on how we can group and present test only functions in the SDK API.

**Review by leighmcculloch — APPROVED** [review]

**Pull request merged**
